tests/test-geometry.lua

```diff
@@ -257,6 +261,134 @@ local steps = {
     end
 }
 
-runner.run_steps(steps)
+-- Testing unmanaged client geometries, which depends on wmctrl as a portable way to clean up unmanaged client.
+if os.getenv('HAS_WMCTRL') == '1' then
+    print("Added tests for unmanaged client geometries.")
+    -- By default titlebar size depends on system font rendering. Make it a fixed value.
+    local fix_titlebar_size = function(c) c:titlebar_top(20) end
+    local test_data = {
+        -- gravity => expectation of unmanaged x, y, w, h.
+        ["NORTH_WEST"] = {0,  0,  100, 80},
+        ["NORTH"] =      {10, 0,  100, 80},
+        ["NORTH_EAST"] = {20, 0,  100, 80},
+        ["WEST"] =       {0,  20, 100, 80},
+        ["CENTER"] =     {10, 20, 100, 80},
+        ["EAST"] =       {20, 20, 100, 80},
+        ["SOUTH_WEST"] = {0,  40, 100, 80},
+        ["SOUTH"] =      {10, 40, 100, 80},
+        ["SOUTH_EAST"] = {20, 40, 100, 80},
+        ["STATIC"] =     {10, 30, 100, 80}
+    }
+    for gravity, expectation in pairs(test_data) do
+        gears.table.merge(steps, {
+        function()
+            set_rules {
+                -- All clients will match this rule.
+                { rule = { },properties = {
+                      titlebars_enabled = true,
+                      border_width      = 10,
+                      border_color      = "#00ff00",
+                      size_hints_honor  = false,
+                      x                 = 0,
+                      y                 = 0,
+                      width             = 100,
+                      height            = 100
+                }}
+            }
+            client.connect_signal("request::titlebars", fix_titlebar_size)
+            -- Wait for the previous cleanup to be done
+            if #client.get() == 0 then
+                return true
+            end
+        end,
+        function(count)
+            if count == 1 then
+                print("testing gravity " .. gravity)
+                test_client(nil,nil,nil,nil,nil,{gravity=lgi.Gdk.Gravity[gravity]})
+            else
+                local c = client.get()[1]
+                if c then
+                    assert(c.size_hints.win_gravity == gravity:lower())
+                    assert(c.border_width      == 10  )
+                    assert(c:geometry().x      == 0   )
+                    assert(c:geometry().y      == 0   )
+                    assert(c:geometry().width  == 100 )
+                    assert(c:geometry().height == 100 )
+                    windowid = c.window
+                    c:unmanage()
```

Comment:
Could you add a call to `awesome.sync()` afterwards? That should make sure that the X11 server *really*  already handled the `c:unmanage()` before we get to the next step.

Reply:
TIL, added.

Reply:
Since you write "TIL", let's actually explain my thinking here:
- `client_unmanage()` in `objects/client.c` "does some stuff" and sends requests to the X11 server.
- However, libxcb actually has an output buffer and does not send requests unless it actually has to.
- Even when the output buffer is flushed, that only means that the requests were sent, but not that they were handled by the X11 server.
- Thus, in theory it could be possible that  `wmctrl` manages to start and talk to the X11 server before the unmanage is actually done / handled by the X11 server.

The call to `awesome.sync()` sends a request to the X11 server and actually waits for the reply. X11 guarantees that the requests are handled in-order. Thus, by waiting for a reply to a new request, we can be sure that all the previous requests were handled.